src/values/ObjectValue.js

```diff
@@ -244,6 +246,10 @@ export default class ObjectValue extends ConcreteValue {
   // to return AbstractValue for unknown properties.
   _isSimple: BooleanValue;
 
+  // If true, it is not safe to perform any more mutations that would change
+  // the object's serialized form.
+  _isFinal: BooleanValue;
```

Comment:
All of the "tracked" properties should include `| AbstractValue` in their type, as merging two different states of the flag might result in an `AbstractValue`. (There's an open issue for that: #1124.)
In that sense, the current handling of `_isPartial` and `_isSimple` is fragile, but at least so far making objects partial or simple is a meta statement that needs to be explicitly inserted by some framework, but your setting of `_isFinal` may happen implicitly as part of the regular program execution flow.

Reply:
Thanks for explaining. I wasn't sure what "tracking" means but this makes more sense now.

Reply:
Does this mean the whole approach is flawed, or just that I should make it accept `AbstractValue` and also treat this as a bailout reason when I check for it?

Reply:
The approach should be fine, it just needs to handle AbstractValues and bail out when it runs into issues.